### PR TITLE
nxp_adsp_imx8m: Clean up and enable (more) tests

### PR DESCRIPTION
**Pull request by iuliana-prodan**
Clean up and enable (more) tests for nxp_adsp_imx8m

### CODE REVIEW

**Comment by iuliana-prodan** [comment]
@LaurentiuM1234 please have a look.

File: boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig
Comment (LaurentiuM1234) [inline boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig:10]: Suggestion: if we know the data cache line size will stay the same for all boards using the XTENSA-based i.MX8M SoC we can treat this as a SoC quirk and move it to the SoC files.

File: dts/xtensa/nxp/nxp_imx8m.dtsi
Comment (LaurentiuM1234) [inline dts/xtensa/nxp/nxp_imx8m.dtsi:70]: Suggestion: use an invalid INTID here (assuming 29 is valid) such that it will result in a build error raised by `gen_isr_tables.py`. My suggestion is kinda bad I guess but at least it may avoid the confusion caused by the fact that the interrupt won't work?

**Review by LaurentiuM1234 — COMMENTED** [review]
LGTM. Just 2 suggestions below.

**Review by dbaluta — DISMISSED** [review]

File: boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig
Comment (iuliana-prodan) [reply boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig:10]: I've kept it here since.. it was already here :)
The stack size is also here.. and other xtensa boards define it here:
```
boards/xtensa/intel_adsp_ace20_lnl/intel_adsp_ace20_lnl_defconfig:16:CONFIG_DCACHE_LINE_SIZE=64
boards/xtensa/intel_adsp_ace15_mtpm/intel_adsp_ace15_mtpm_defconfig:16:CONFIG_DCACHE_LINE_SIZE=64
boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_tgph_defconfig:24:CONFIG_DCACHE_LINE_SIZE=64
boards/xtensa/intel_adsp_cavs25/intel_adsp_cavs25_defconfig:23:CONFIG_DCACHE_LINE_SIZE=64
boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig:10:CONFIG_DCACHE_LINE_SIZE=128
boards/arm64/mimx93_evk/mimx93_evk_a55_sof_defconfig:25:CONFIG_DCACHE_LINE_SIZE=64
```

I realize now that for `nxp_adsp_imx8` and `nxp_adsp_imx8x` is in soc files, so I can move it there for imx8m also, to be consisted on `nxp_adsp`.

File: dts/xtensa/nxp/nxp_imx8m.dtsi
Comment (iuliana-prodan) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: I've tried to use an invalid value like 33, for example, but, when compiling some uart or shell samples (for testing), I'm getting:
```
FAILED: zephyr/isr_tables.c /opt/samba/nxa06898/zephyrproject/zephyr/twister-out/nxp_adsp_imx8m/samples/drivers/uart/echo_bot/sample.drivers.uart/zephyr/isr_tables.c
cd /opt/samba/nxa06898/zephyrproject/zephyr/twister-out/nxp_adsp_imx8m/samples/drivers/uart/echo_bot/sample.drivers.uart/zephyr && /usr/bin/python /opt/samba/nxa06898/zephyrproject/zephyr/scripts/build/gen_isr_tables.py --output-source isr_tables.c --kernel /opt/samba/nxa06898/zephyrproject/zephyr/twister-out/nxp_adsp_imx8m/samples/drivers/uart/echo_bot/sample.drivers.uart/zephyr/zephyr_pre0.elf --intlist-section .intList --intlist-section intList --sw-isr-table
gen_isr_tables.py: error: IRQ 33 (offset=0) exceeds the maximum of 31
```
So, I hope the comment is enough to not try to use UART interrupt-driven

File: boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig
Comment (LaurentiuM1234) [reply boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig:10]: ACK, up to you.

Comment (iuliana-prodan) [reply boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig]: I've sent a v2 where I changed the DCACHE_LINE_SIZE to soc/.

**Review by LaurentiuM1234 — APPROVED** [review]

File: dts/xtensa/nxp/nxp_imx8m.dtsi
Comment (dcpleung) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: Could you not add this until there is proper UART interrupt support? It should be possible to have UART as polling only. If the driver requires this, that is not right. The driver should ignore the property if it's not present.

Comment (iuliana-prodan) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: UART in polling mode is functional now.
But, for some tests ( or samples to compile) I had to add this.
I'm getting:
```
zephyrproject/zephyr/twister-out/nxp_adsp_imx8m/samples/subsys/shell/devmem_load/sample.devmem_load.uart.interrupt/zephyr/include/generated/devicetree_generated.h:1436:35: error: 'DT_N_S_soc_S_uart_30a60000_IRQ_IDX_0_VAL_irq' undeclared (first use in this function); did you mean 'DT_N_S_soc_S_uart_30a60000_REG_IDX_0_VAL_SIZE'?
 1436 | #define DT_N_INST_0_nxp_imx_iuart DT_N_S_soc_S_uart_30a60000
      |                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

I'll try to find a proper fix.

Comment (dcpleung) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: I guess the issue is the mixing of polling only and interrupt driven UARTs on the same system. The proper fix would be to amend `UART_IMX_CONFIG_FUNC()` and `UART_IMX_IRQ_CFG_FUNC_INIT()` in the UART driver to be conditional on whether the interrupt property is there.

Comment (dcpleung) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: BTW, how long do you think it will take to actually enable interrupt? If it's soon enough, maybe we can just merge this now as-is, and skip modifying the driver.

Comment (iuliana-prodan) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: @dcpleung I've look more in depth how UART interrupt driven work and, for i.MX8MP (and other targets) we have an IRQ_Steer interrupt controller that redirects the incoming interrupts, from peripherals to output interrupts on a designated channel.
My colleague @LaurentiuM1234  is working on this, actually this is already in review: https://github.com/zephyrproject-rtos/zephyr/pull/62776

So, in order to have the UART working with interrupts will need the IRQ_Steer PR merged.

Next, I'll have to do some updates in the dtsi like:
```
diff --git a/dts/xtensa/nxp/nxp_imx8m.dtsi b/dts/xtensa/nxp/nxp_imx8m.dtsi
index 8c00fa47de..f953994d32 100644
--- a/dts/xtensa/nxp/nxp_imx8m.dtsi
+++ b/dts/xtensa/nxp/nxp_imx8m.dtsi
@@ -43,6 +43,21 @@
        };
 
        soc {
+               irqsteer: interrupt-controller@30a80000 {
+                               compatible = "nxp,irqsteer-intc";
+                               reg = <0x30a80000 DT_SIZE_K(64)>;
+                               #size-cells = <0>;
+                               #address-cells = <1>;
+
+                               master0: interrupt-controller@0 {
+                                       compatible = "nxp,irqsteer-master";
+                                       reg = <0>;
+                                       interrupt-controller;
+                                       #interrupt-cells = <1>;
+                                       interrupts-extended = <&clic 19 0 0>;
+                               };
+               };
+
                ccm: ccm@30380000 {
                        compatible = "nxp,imx-ccm";
                        reg = <0x30380000 DT_SIZE_K(64)>;
@@ -66,10 +81,8 @@
                uart4: uart@30a60000 {
                        compatible = "nxp,imx-iuart";
                        reg = <0x30a60000 0x10000>;
-                       /* TODO: This INTID is just a dummy
-                        * until we can support UART interrupts
-                        */
-                       interrupts = <29 0>;
+                       interrupt-parent = <&master0>;
+                       interrupts = <29>;
                        clocks = <&ccm IMX_CCM_UART4_CLK 0x6c 24>;
                        status = "disabled";
                };
```
and some configs enabled:
```
diff --git a/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig b/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig
index 72ccd09f55..98eaa53016 100644
--- a/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig
+++ b/boards/xtensa/nxp_adsp_imx8m/nxp_adsp_imx8m_defconfig
@@ -27,3 +27,13 @@ CONFIG_UART_CONSOLE=y
 
 # enable pin controller
 CONFIG_PINCTRL=y
+
+# interrupt-related configurations
+CONFIG_MULTI_LEVEL_INTERRUPTS=y
+CONFIG_2ND_LEVEL_INTERRUPTS=y
+CONFIG_2ND_LVL_ISR_TBL_OFFSET=32
+CONFIG_MAX_IRQ_PER_AGGREGATOR=64
+CONFIG_NUM_2ND_LEVEL_AGGREGATORS=3
+CONFIG_2ND_LVL_INTR_00_OFFSET=19
+CONFIG_2ND_LVL_INTR_01_OFFSET=20
+CONFIG_2ND_LVL_INTR_02_OFFSET=21
```

Therefore, we can wait for the IRQ_Steer to be merged and I'll update _this PR_ after that, or we can merge this as-is and follow-up with a _new PR_ to enable interrupts for UART.
I prefer second option.


cc: @dbaluta @LaurentiuM1234

Comment (LaurentiuM1234) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: My 2 cents here regarding this issue:

We can't enable the IRQ_STEER driver on any of the i.MX8 boards w/o fully switching to the Zephyr native drivers  as this will lead to some awkward scenarios on SOF side.

The above sample seems to test interrupt-driven UART, which will indeed require the `interrupts` property be specified (which does make sense since, well, you can't have interrupt-driven UART w/o an interrupt). So I see 3 choices here:

1. Don't extend the test set or find a way to skip those using interrupt-driven UART.
2. Merge PR as-is. The dummy interrupt issue will be addressed later on when IRQ_STEER is enabled so the dummy INTID won't stay there forever.
3. If choice 2) isn't OK then use the DSP direct interrupt which is asserted by the UART interrupt. This isn't the best choice either but at least we're using the actual level 1 INTID. Of course, this will break w/o shared interrupt support if there's multiple nodes using the same strategy.

Personally, I prefer choice 2 since the problem will be addressed later on and this was also done on QM/QXP, but I do understand that this isn't really the _right_ choice.

Comment (dcpleung) [reply dts/xtensa/nxp/nxp_imx8m.dtsi:70]: Since there is a path forward, I guess it is okay to merge as-is.

**Review by dcpleung — APPROVED** [review]

**Review by dbaluta — APPROVED** [review]

**Comment by dbaluta** [comment]
@iuliana-prodan Thanks! This looks good to me and is ready to be merged.

**Pull request merged**
